### DISCUSSION TITLE
Any particular reason why localAI doesn't parallelize request processing?

### DISCUSSION THREAD

**Discussion by ghevge** [mudler/LocalAI#1670] [General]
I have deploy  the latest localAI container (v2.7.0) on a 2 x Xeon 2680V4 (56 threads in total) with 198 GB of RAM, but from what I can tell, the request hitting the /v1/chat/completions endpoint, are being processed one after the other, not in parallel. 
I believe there are enough resources on my system to process these requests in parallel.

I am using the openhermes-2.5-mistral-7b.Q8_0.gguf model.

To get a response for the curl example below, it takes about 20+ seconds. Is this good or bad for the system I have?

Thanks

PS: I can't install a GPU on this system as it is a 1U unit. 

Curl: 

```
curl http://myhost:8082/v1/chat/completions  -H "Content-Type: application/json"  -d '{
 "model": "openhermes-2.5-mistral-7b.Q8_0.gguf",
 "response_format": { "type": "json_object" },
 "max_tokens": 50,
 "temperature": 0.7,
 "top_p": 0.4,
 "top_k": 40,
 "n": 1,
 "messages": [
        {"role": "system", "content": "You are an AI assistant who gives a quality response to whatever humans ask of you."},
         {"role": "user", "content": "### Human:\n list only tags from text, comma separated: may continue tocopherols or tocopheryl type of vitamin e derived from vegetable oil soy derivative \n ### Assistant:"}] 
 }'
```

Docker-compose:

```
version: '3.8'

services:
  localai:
    image: 192.168.1.95:81/localai:latest
    ports:
      - 8082:8080
    env_file:
      - env-localai.txt
    volumes:
      - /datanvme/models:/models:cached
    command: ["/usr/bin/local-ai" ]
    networks:
      mw-network:
        aliases:
          - localai

networks:
  mw-network:
    external:
      name: mw-network
```
LocalAI logs:

```
Recreating docker_localai_1 ... done
Attaching to docker_localai_1
localai_1  | @@@@@
localai_1  | Skipping rebuild
localai_1  | @@@@@
localai_1  | If you are experiencing issues with the pre-compiled builds, try setting REBUILD=true
localai_1  | If you are still experiencing issues with the build, try setting CMAKE_ARGS and disable the instructions set as needed:
localai_1  | CMAKE_ARGS="-DLLAMA_F16C=OFF -DLLAMA_AVX512=OFF -DLLAMA_AVX2=OFF -DLLAMA_FMA=OFF"
localai_1  | see the documentation at: https://localai.io/basics/build/index.html
localai_1  | Note: See also https://github.com/go-skynet/LocalAI/issues/288
localai_1  | @@@@@
localai_1  | CPU info:
localai_1  | model name : Intel(R) Xeon(R) CPU E5-2680 v4 @ 2.40GHz
localai_1  | flags              : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single pti ssbd ibrs ibpb stibp tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm rdt_a rdseed adx smap intel_pt xsaveopt cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts vnmi md_clear flush_l1d
localai_1  | CPU:    AVX    found OK
localai_1  | CPU:    AVX2   found OK
localai_1  | CPU: no AVX512 found
localai_1  | @@@@@
localai_1  | 4:09PM DBG no galleries to load
localai_1  | 4:09PM INF Starting LocalAI using 4 threads, with models path: /models
localai_1  | 4:09PM INF LocalAI version: v2.7.0 (abd678e147e29ab46a038d28c4806768b2b7cde5)
localai_1  | 4:09PM WRN [startup] failed resolving model '/usr/bin/local-ai'
localai_1  | 4:09PM INF Preloading models from /models
localai_1  |
localai_1  |  ┌───────────────────────────────────────────────────┐
localai_1  |  │                   Fiber v2.50.0                   │
localai_1  |  │               http://127.0.0.1:8080               │
localai_1  |  │       (bound on host 0.0.0.0 and port 8080)       │
localai_1  |  │                                                   │
localai_1  |  │ Handlers ............ 73  Processes ........... 1 │
localai_1  |  │ Prefork ....... Disabled  PID ................ 14 │
localai_1  |  └───────────────────────────────────────────────────┘
localai_1  |

```

**Comment by mudler** [comment]
With CPU is expected to have a slower inference time, but you can enable parallel requests by setting the environment variable appropriately: https://github.com/mudler/LocalAI/blob/39a6b562cfb1a8c488fc2560e4f32abebfa13d8e/.env#L72

However that would probably not work very well with CPUs.

**Comment by pmarini-nc** [comment]
Hello @mudler, sorry to bring this up on an old thread, but it is not clear to me whether `LLAMACPP_PARALLEL` indicates the number of requests that the system can execute concurrently? My use case is using LocalAI as a backend for the Nextcloud Assistant via [LocalAI integration](https://github.com/nextcloud/integration_openai). I'm using CPUs and the idea is to share the LocalAI engine across several Nextcloud instances, or from the same instance allowing simultaneous task executions from several users.

Thanks!

**Comment by pmarini-nc** [comment]
Hello,

So it seems that I cannot get parallel requests to work.

The use case is to run two inference tasks in parallel (simulating for example, two users asking to summarize a text from Nextcloud Assistant - see my previous comment).

I pass the following environment variables to the LocalAI runtime:

```
LOCALAI_THREADS=2
PARALLEL_REQUESTS=true
LLAMACPP_PARALLEL=2
```

During the test I see just one process capping at 200% CPU in top:
`
/etc/local-ai/backends/cpu-llama-cpp/lib/ld.so /etc/local-ai/backends/cpu-llama-cpp/llama-cpp-avx2 --addr 127.0.0.1:36023`

So this means that only one request is executing at a given time. What should I do to run the tasks in parallel? Should I expect to see 2 ld.so processes each one runing on $LOCALAI_THREADS threads? Is there a way to monitor the running tasks?

It would be great to have your feedback, @mudler !

Thanks in advance!

**Comment by scetu** [comment]
Hey, I have found the solution, based on following line in grcp-server
https://github.com/mudler/LocalAI/blob/d4cd6c284f055aa61cb6a6c67242fe1cb4456d1f/backend/cpp/llama-cpp/grpc-server.cpp#L444

you need to add into your model config `n_parallel:2`
```
backend: llama-cpp
options:
    - use_jinja:true
    - n_parallel:2
```
